codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -137,145 +207,115 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-            childChannel.streamId(streamId);
-        } else {
-            ChannelFuture future = bootstrap.connect(streamId);
-            childChannel = (Http2StreamChannel) future.channel();
+        // TODO: Should this happen now that onStreamActive(...) is called when an Http2FrameStreamEvent with state
+        // ACTIVE is received.
+        if (childChannel == null) {
+            childChannel = onStreamActive(stream);
         }
 
-        Http2StreamChannel existing = childChannels.put(streamId, childChannel);
-        assert existing == null;
+        fireChildReadAndRegister(childChannel, frame);
     }
 
-    private void onStreamClosed(int streamId) {
-        final Http2StreamChannel childChannel = childChannels.remove(streamId);
+    private void onStreamClosed(Http2FrameStream stream) {
+        Http2StreamChannel childChannel = channels.get(stream);
         if (childChannel != null) {
-            final EventLoop eventLoop = childChannel.eventLoop();
-            if (eventLoop.inEventLoop()) {
-                onStreamClosed0(childChannel);
-            } else {
-                eventLoop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        onStreamClosed0(childChannel);
-                    }
-                });
-            }
+            childChannel.streamClosed();
         }
     }
 
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
+    private Http2StreamChannel onStreamActive(Http2FrameStream stream) {
+        Http2StreamChannel childChannel = channels.get(stream);
+        if (childChannel == null) {
+            childChannel = newStreamChannel(stream);
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+            childChannel.pipeline().addLast(inboundStreamHandler);
+
+            ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+            future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
+        }
+
+        assert !childChannel.isWritable();
+        childChannel.incrementOutboundFlowControlWindow(initialOutboundStreamWindow);
+        childChannel.pipeline().fireChannelWritabilityChanged();
+        return childChannel;
     }
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
-        } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)